Expose `ProposalRelevanceCfg` in user preferences

mlangc opened the pull request:
Make the recently introduced `ProposalRelevanceCfg` (see [ticket 1002686](https://app.assembla.com/spaces/scala-ide/tickets/1002686-scala-type-completions-should-be-sorted-differently/details#)) accessible in the UI.

I know that using full regular expressions for package names is a bit cumbersome, but it allows maximum flexibility.

sschaef commented on org.scala-ide.sdt.core/plugin.xml:
```diff
@@ -897,6 +903,9 @@
            class="org.scalaide.ui.internal.preferences.AutoEditsPreferenceInitializer">
      </initializer>
+     <initializer
+           class="org.scalaide.ui.internal.preferences.ComplitionPreferenceInitializer">
```
`s/Complition/Completion/`

sschaef commented on org.scala-ide.sdt.core/src/org/scalaide/core/completion/ProposalRelevanceCfg.scala:
```diff
@@ -29,3 +35,22 @@ object DefaultProposalRelevanceCfg extends ProposalRelevanceCfg {
 }
+
+object PrefsBasedProposalRelevanceCfg extends ProposalRelevanceCfg {
+  val PFavoritePackages = "scala.tools.eclipse.ui.preferences.completions.favoritePackages"
```
Please move everything that has to do with the pref store and all of these constants to `CompletionPreferencePage` because that is the location where we usually keep this stuff.

mlangc replied:
I've refactored the code a bit... I hope it is more to your liking now ;-)

sschaef commented on org.scala-ide.sdt.core/src/org/scalaide/ui/internal/preferences/CompletionPreferencePage.scala:
```diff
@@ -0,0 +1,99 @@
+  }
+}
+
+class ComplitionPreferenceInitializer extends AbstractPreferenceInitializer {
```
`s/Complition/Completion/`

mlangc replied:
Thanks for spotting... should be fixed now!

sschaef commented on org.scala-ide.sdt.core/src/org/scalaide/ui/internal/preferences/CompletionPreferencePage.scala:
```diff
@@ -0,0 +1,99 @@
+      store.setDefault(key, StringListSerializer.serialize(regexes.map(_.regex)))
+    }
+
+    setDefault(PFavoritePackages, DefaultProposalRelevanceCfg.favoritePackages)
```
`favoritePackages` is read from the pref store, but the default preferences can't be read from the pref store because the defaults is what exists first.

mlangc replied:
Hmm, the defaults are not read from the preferences, but are hard coded into `org.scalaide.core.completion.DefaultProposalRelevanceCfg`...

wpopielarski commented on org.scala-ide.sdt.core.tests/src/org/scalaide/ui/internal/preferences/StringListSerializerTest.scala:
```diff
@@ -0,0 +1,62 @@
+
+  @Test
+  def testWithIllegalInput(): Unit = {
+    expectExceptionOnDeserialize("1")
```
I don't get this test. Why is the `"1"` not valid if is valid in `singleString()` ?

mlangc replied:
Because the format is `n1|n1Charsn2|n2Chars....nk|nkChars`. A number `n` therefore must always be followed by `|` and `n` characters. This is even true for the empty string, that is encoded as `0|`. Is this more understandable now? Do you want me to document the format in the Scaladocs?

sschaef replied:
I would like to see the format documented in Scaladoc.

wpopielarski commented on org.scala-ide.sdt.core.tests/src/org/scalaide/ui/internal/preferences/StringListSerializerTest.scala:
```diff
@@ -0,0 +1,62 @@
+  private def testWith(strs: String*): Unit = {
+    val serialized = StringListSerializer.serialize(strs)
+    val deserialized = StringListSerializer.deserialize(serialized)
+    assertEquals(s"Error related to sererialized representation '$serialized'", strs, deserialized)
```
`s/sererialized/deserialized`

mlangc commented:
@sschaef , @wpopielarski I've just refactored the code around `StringListSerializer`, that is now called `StringListMapper` a bit, and also added some Scaladocs. In them I also state that the exact encoding, (that is much clearer described in Scala than in English imo) should be regarded as an implementation detail.

sschaef commented:
LGTM

One thing I'm not satisfied yet is that it is not obvious how to understand the configuration options in the new preference page. It should tell the user that these options are responsible for the ordering of the code completion entries and how exactly it affects the ordering. But this can't really be described in the preferences page. Some code examples or screenshots on the homepage may be best to describe the behavior. @mlangc If you you interested in writing such documentation, fell welcome to do it. Otherwise I may make it by myself before we do a release.

mlangc commented:
@sschaef you think about adding some documentation [here](https://github.com/scala-ide/docs/blob/master/src/sphinx/4.0.x/features/typingviewing/code-completion.rst)? I would happily do that....

sschaef commented:
Yes, it would be great to put it to the code completion section.

mlangc commented:
See https://github.com/scala-ide/docs/pull/144.
